pyrdp/parser/rdp/ntlmssp.py

```diff
@@ -86,3 +92,109 @@ def parseNTLMSSPAuthenticate(self, data: bytes, stream: BytesIO) -> NTLMSSPAuthe
+        ber.readLength(stream)
+        ber.readContextualTag(stream, 0, True)
+        
+        negoTokens = BytesIO(ber.readOctetString(stream))            # NegoData
+        return NTLMSSPTSRequestPDU(version, negoTokens)
+
+    def parseNTLMSSPChallengePayload(self, data: bytes, stream: BytesIO, workstationLen: int) -> NTLMSSPChallengePayloadPDU:
+        stream.read(workstationLen)
+        return NTLMSSPChallengePayloadPDU(workstation)
+
+    def writeNTLMSSPChallenge(self, workstation: str, serverChallenge: bytes) -> bytes:
+        stream = BytesIO()
+        substream = BytesIO()
+        
+        workstation = workstation.encode('utf-16le')
+        nameLen = len(workstation)
+        pairsLen = self.writeNTLMSSPChallengePayload(substream, workstation)
+        
+        substream.write(b'NTLMSSP\x00')                                                    # signature
+        substream.write(Uint32LE.pack(NTLMSSPMessageType.CHALLENGE_MESSAGE))               # message type
+        substream.write(Uint16LE.pack(nameLen))                                            # workstation length
+        substream.write(Uint16LE.pack(nameLen))                                            # workstation max length
+        substream.write(Uint32LE.pack(NTLMSSPChallengeType.WORKSTATION_BUFFER_OFFSET))     # workstation buffer offset
+        substream.write(Uint32LE.pack(NTLMSSPChallengeType.NEGOTIATE_FLAGS))               # negotiate flags
+        substream.write(serverChallenge)                                                   # server challenge
+        substream.write(Uint64LE.pack(0))                                                  # reserved
+        substream.write(Uint16LE.pack(pairsLen))                                           # target info len
+        substream.write(Uint16LE.pack(pairsLen))                                           # target info max len
+        substream.write(Uint32LE.pack(NTLMSSPChallengeType.WORKSTATION_BUFFER_OFFSET + nameLen)) # target info buffer offset
+        substream.write(Uint8.pack(NTLMSSPChallengeVersion.NEG_PROD_MAJOR_VERSION_HIGH))   # product major version
+        substream.write(Uint8.pack(NTLMSSPChallengeVersion.NEG_PROD_MINOR_VERSION_LOW))    # product minor version
+        substream.write(Uint16LE.pack(NTLMSSPChallengeVersion.NEG_PROD_VERSION_BUILT))     # product build
+        substream.write(Uint8.pack(0))                                                     # reserved
+        substream.write(Uint8.pack(0))                                                     # reserved
+        substream.write(Uint8.pack(0))                                                     # reserved
+        substream.write(Uint8.pack(NTLMSSPChallengeVersion.NEG_NTLM_REVISION_CURRENT))     # NTLM revision current
+        
+        self.writeNTLMSSPTSRequest(stream, NTLMSSPChallengeVersion.CREDSSP_VERSION, substream.getvalue())
+        return stream.getvalue()
```

Comment:
I'm going to close this review because there is already a couple of items to fix and look at but I wonder if we can simplify this pattern somehow. We should check for other code examples where long sequences of bytes are written like that.

Reply:
Should I resolve this?